pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -962,17 +974,41 @@ func (t *Test) chooseUpgradePath() ([]*clusterupgrade.Version, error) {
 			return []*clusterupgrade.Version{pred}, nil
 		}
 
-		predPred, err := t.options.predecessorFunc(t.prng, pred, t.options.minimumSupportedVersion)
+		// If the predecessor same series as the minimum supported version,
+		// we need to check that it isn't the only possible upgrade where
+		// user steps can run. If it is, don't allow a skip upgrade or else
+		// we will generate a plan with zero upgrades that are able to run user hooks.
+		//
+		// Much of the mixed version framework assumes that there is at least one upgrade
+		// that can run user hooks. We just disable skip version upgrades in this case since
+		// it doesn't make sense to allow a test plan with no user hooks run.
+		if pred.Series() == t.options.minimumSupportedVersion.Series() {
+			maxTestUpgrades, err := release.MajorReleasesBetween(&t.options.minimumSupportedVersion.Version, &currentVersion.Version)
+			if err != nil {
+				return nil, err
+			}
+			if maxTestUpgrades == 1 {
```

Comment:
What precondition ensures it's not zero?

Reply:
`assertValidTest` will throw a failure if it isn't:

```
// Validate that the minimum bootstrap version if set.
	if minBootstrapVersion := test.options.minimumBootstrapVersion; minBootstrapVersion != nil {
		// The minimum bootstrap version should be from an older major version.
		validVersion = minBootstrapVersion.Major().LessThan(currentVersion.Major())
		if !validVersion {
			fail(
				fmt.Errorf(
					"invalid test options: minimum bootstrap version (%s) should be from an older release series than current version (%s)",
					minBootstrapVersion.Version.String(), currentVersion.Version.String(),
				),
			)
```

Reply:
Changed the check in `assertValidTest` to make sure we get at least one upgrade. Interestingly it seemed from the failure message and test case this was always the intention, but the check only made sure the msv was strictly less than the current version.